crates/aptos-admin-service/src/server/consensus/mod.rs

```diff
@@ -0,0 +1,217 @@
+    info!("Dumping quorum store db.");
+
+    match spawn_blocking(move || dump_quorum_store_db(quorum_store_db.as_ref())).await {
+        Ok(result) => {
+            info!("Finished dumping quorum store db.");
+            let headers: Vec<(_, HeaderValue)> =
+                vec![(CONTENT_LENGTH, HeaderValue::from(result.len()))];
+            Ok(reply_with(headers, result))
+        },
+        Err(e) => {
+            info!("Failed to dump quorum store db: {e:?}");
+            Ok(reply_with_status(
+                StatusCode::INTERNAL_SERVER_ERROR,
+                e.to_string(),
+            ))
+        },
+    }
+}
+
+pub async fn handle_dump_block_request(
+    req: Request<Body>,
+    consensus_db: Arc<dyn PersistentLivenessStorage>,
+    quorum_store_db: Arc<dyn QuorumStoreStorage>,
+) -> hyper::Result<Response<Body>> {
+    // TODO(grao): Support bcs encoding.
+    let query = req.uri().query().unwrap_or("");
+    let query_pairs: HashMap<_, _> = url::form_urlencoded::parse(query.as_bytes()).collect();
+
+    let block_id: Option<HashValue> = match query_pairs.get("block_id") {
+        Some(val) => match val.parse() {
+            Ok(val) => Some(val),
+            Err(err) => return Ok(reply_with_status(StatusCode::BAD_REQUEST, err.to_string())),
+        },
+        None => None,
+    };
+
+    if let Some(block_id) = block_id {
+        info!("Dumping block ({block_id:?}).");
+    } else {
+        info!("Dumping all blocks.");
+    }
+
+    match spawn_blocking(move || {
+        dump_blocks(consensus_db.as_ref(), quorum_store_db.as_ref(), block_id)
+    })
+    .await
+    {
+        Ok(result) => {
+            info!("Finished dumping block(s).");
+            let headers: Vec<(_, HeaderValue)> =
+                vec![(CONTENT_LENGTH, HeaderValue::from(result.len()))];
+            Ok(reply_with(headers, result))
+        },
+        Err(e) => {
+            info!("Failed to dump block(s): {e:?}");
+            Ok(reply_with_status(
+                StatusCode::INTERNAL_SERVER_ERROR,
+                e.to_string(),
+            ))
+        },
+    }
+}
+
+fn dump_consensus_db(consensus_db: &dyn PersistentLivenessStorage) -> anyhow::Result<String> {
+    let mut body = String::new();
+
+    let (last_vote, highest_tc, consensus_blocks, consensus_qcs) =
+        consensus_db.consensus_db().get_data()?;
+
+    body.push_str(&format!("Last vote: \n{last_vote:?}\n\n"));
+    body.push_str(&format!("Highest tc: \n{highest_tc:?}\n\n"));
+    body.push_str(&format!("Blocks: \n"));
+    for block in consensus_blocks {
+        body.push_str(&format!(
+            "[id: {:?}, author: {:?}, epoch: {}, round: {:02}, parent_id: {:?}, timestamp: {}, payload: {:?}]\n\n",
+            block.id(),
+            block.author(),
+            block.epoch(),
+            block.round(),
+            block.parent_id(),
+            block.timestamp_usecs(),
+            block.payload(),
+        ));
+    }
+    body.push_str(&format!("QCs: \n"));
```

Comment:
and why not output the whole thing as Json (or json per line), to enable processing with jq / more sophisticated scripts

Reply:
ditto

Reply:
Can you clarify how we plan to consume this? If this is needed for some other tooling (like replay tool), then dumping this as binary file might be useful?

Reply:
I expect only the block?block_id=xxx API will be consumed by program (I left a TODO to support bcs format for that). For other API I only expect people use it in browser so I'm trying to make them human friendly.